src/Perlang.Interpreter/PerlangInterpreter.cs

```diff
@@ -1080,7 +1090,10 @@ public VoidObject VisitWhileStmt(Stmt.While stmt)
                     {
                         // Both values are some form of integers. The BigInteger implementation is more likely to yield
                         // a result that is useful for us in this case.
-                        return BigInteger.Pow(leftNumber, rightNumber);
+                        //
+                        // Because of "Perlang.NullObject is not comparable" checks in TypeResolver, we can assert that
+                        // neither one of these can be `null` at this point.
+                        return BigInteger.Pow((BigInteger)leftNumber!, (int)rightNumber!);
```

Comment:
Because of the more aggressive trimming in .NET 6, we ran into this problem now. This manual casting works around it, since it makes the .NET compiler be more aware of the exact method being invoked here.

The long-term plan would be to get rid of `dynamic` altogether.

```
$ make release docs-test-examples 
dotnet publish src/Perlang.ConsoleApp/Perlang.ConsoleApp.csproj -c Release -r linux-x64 --self-contained true /p:PublishReadyToRun=true /p:SolutionDir=$(pwd)/
Microsoft (R) Build Engine version 17.0.0-preview-21501-01+bbcce1dff for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored /home/per/git/perlang/src/Perlang.ConsoleApp/Perlang.ConsoleApp.csproj (in 316 ms).
  Restored /home/per/git/perlang/src/Perlang.Common/Perlang.Common.csproj (in 316 ms).
  Restored /home/per/git/perlang/src/Perlang.Parser/Perlang.Parser.csproj (in 316 ms).
  Restored /home/per/git/perlang/src/Perlang.Stdlib/Perlang.Stdlib.csproj (in 316 ms).
  Restored /home/per/git/perlang/src/Perlang.Interpreter/Perlang.Interpreter.csproj (in 316 ms).
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  Perlang.Common -> /home/per/git/perlang/src/Perlang.Common/bin/Release/net6.0/Perlang.Common.dll
  Perlang.Parser -> /home/per/git/perlang/src/Perlang.Parser/bin/Release/net6.0/Perlang.Parser.dll
  Perlang.Stdlib -> /home/per/git/perlang/src/Perlang.Stdlib/bin/Release/net6.0/Perlang.Stdlib.dll
  Perlang.Interpreter -> /home/per/git/perlang/src/Perlang.Interpreter/bin/Release/net6.0/Perlang.Interpreter.dll
  Perlang.ConsoleApp -> /home/per/git/perlang/src/Perlang.ConsoleApp/bin/Release/net6.0/linux-x64/perlang.dll
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  Perlang.ConsoleApp -> /home/per/git/perlang/src/Perlang.ConsoleApp/bin/Release/net6.0/linux-x64/publish/
for e in docs/examples/*.per ; do echo == $e ; src/Perlang.ConsoleApp/bin/Release/net6.0/linux-x64/publish/perlang $e ; echo ; done
== docs/examples/hello_world.per
Hello World

== docs/examples/pi.per
Unhandled exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> Microsoft.CSharp.RuntimeBinder.RuntimeBinderException: 'System.Numerics.BigInteger' does not contain a definition for 'Pow'
   at CallSite.Target(Closure , CallSite , Type , BigInteger , Object )
   at System.Dynamic.UpdateDelegates.UpdateAndExecute3[T0,T1,T2,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2)
   at Perlang.Interpreter.PerlangInterpreter.VisitBinaryExpr(Binary expr) in /home/per/git/perlang/src/Perlang.Interpreter/PerlangInterpreter.cs:line 1096
   at Perlang.Expr.Binary.Accept[TR](IVisitor`1 visitor) in /home/per/git/perlang/src/Perlang.Common/Expr.cs:line 101
   at Perlang.Interpreter.PerlangInterpreter.Evaluate(Expr expr) in /home/per/git/perlang/src/Perlang.Interpreter/PerlangInterpreter.cs:line 836
   at Perlang.Interpreter.PerlangInterpreter.VisitGroupingExpr(Grouping expr) in /home/per/git/perlang/src/Perlang.Interpreter/PerlangInterpreter.cs:line 821
   at Perlang.Expr.Grouping.Accept[TR](IVisitor`1 visitor) in /home/per/git/perlang/src/Perlang.Common/Expr.cs:line 175
   at Perlang.Interpreter.PerlangInterpreter.Evaluate(Expr expr) in /home/per/git/perlang/src/Perlang.Interpreter/PerlangInterpreter.cs:line 836
   at Perlang.Interpreter.PerlangInterpreter.VisitBinaryExpr(Binary expr) in /home/per/git/perlang/src/Perlang.Interpreter/PerlangInterpreter.cs:line 1000
   at Perlang.Expr.Binary.Accept[TR](IVisitor`1 visitor) in /home/per/git/perlang/src/Perlang.Common/Expr.cs:line 101
   at Perlang.Interpreter.PerlangInterpreter.Evaluate(Expr expr) in /home/per/git/perlang/src/Perlang.Interpreter/PerlangInterpreter.cs:line 836
   at Perlang.Interpreter.PerlangInterpreter.VisitVarStmt(Var stmt) in /home/per/git/perlang/src/Perlang.Interpreter/PerlangInterpreter.cs:line 949
   at Perlang.Stmt.Var.Accept[TR](IVisitor`1 visitor) in /home/per/git/perlang/src/Perlang.Common/Stmt.cs:line 166
   at Perlang.Interpreter.PerlangInterpreter.Execute(Stmt stmt) in /home/per/git/perlang/src/Perlang.Interpreter/PerlangInterpreter.cs:line 858
   at Perlang.Interpreter.PerlangInterpreter.Interpret(IEnumerable`1 statements) in /home/per/git/perlang/src/Perlang.Interpreter/PerlangInterpreter.cs:line 553
   at Perlang.Interpreter.PerlangInterpreter.Eval(String source, ScanErrorHandler scanErrorHandler, ParseErrorHandler parseErrorHandler, NameResolutionErrorHandler nameResolutionErrorHandler, ValidationErrorHandler typeValidationErrorHandler, ValidationErrorHandler immutabilityValidationErrorHandler, CompilerWarningHandler compilerWarningHandler) in /home/per/git/perlang/src/Perlang.Interpreter/PerlangInterpreter.cs:line 280
   at Perlang.ConsoleApp.Program.Run(String source, CompilerWarningHandler compilerWarningHandler) in /home/per/git/perlang/src/Perlang.ConsoleApp/Program.cs:line 352
   at Perlang.ConsoleApp.Program.RunFile(String path) in /home/per/git/perlang/src/Perlang.ConsoleApp/Program.cs:line 312
   at Perlang.ConsoleApp.Program.<>c__DisplayClass9_0.<MainWithCustomConsole>b__3(ParseResult parseResult, IConsole console) in /home/per/git/perlang/src/Perlang.ConsoleApp/Program.cs:line 195
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object , Span`1& , Signature , Boolean , Boolean )
   at System.Reflection.RuntimeMethodInfo.Invoke(Object , BindingFlags , Binder , Object[] , CultureInfo )
   at System.Delegate.DynamicInvokeImpl(Object[] args)
   at System.Delegate.DynamicInvoke(Object[] args)
   at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__24_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseDebugDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

Reply:
#225 has been resolved. :tada: Resolving this conversation as well; it can always be expanded if we are interested in the history details.